<a id="cont"></a>

## Table of Contents

<a href=#one>1. Project Overview</a>

<a href=#two>2. Import Packages and Loading Data</a>

<a href=#three>3. Exploring Data </a>

<a href=#four>4. Exploratory Data Analysis (EDA)</a>

<a href=#five>5. Pre-Processing and Cleaning Data</a>

<a href=#six>6. Modeling</a>

<a href=#seven>7. Model Evaluation and Conclusions</a>

 <a id="one"></a>
## 1. Project Overview


**Objective:** In today's competitive landscape, effective sales management is crucial for optimizing distribution channels, reducing costs, and maximizing profits.

This project focused on analyzing sales and revenue trends to provide data-driven insights for informed decision-making. Through data analysis, visualization, and the development of a comprehensive dashboard, we aimed to enhance the company's ability to identify key markets and improve overall sales performance and revenue growth.

### **Problem Statement**

As competition intensifies and businesses strive to optimize operations, managing sales effectively has become more critical than ever. Companies need to identify key markets and understand sales performance trends to remain competitive and drive revenue growth. However, without clear insights into sales and revenue data, decision-makers face challenges in optimizing distribution, reducing costs, and increasing profitability. This project seeks to address these challenges by leveraging data analysis and visualization to uncover actionable insights, enabling more informed, data-driven decisions for improving sales and revenue performance.


### **Sales Transactions Data Summary (2019)**

This notebook provides an overview of a sales transaction dataset from an online electronics store. The data contains orders placed in 2019, including product details, prices, and customer purchase addresses.

#### **1. Dataset Description**

The dataset consists of the following columns:
- **Order ID**: A unique identifier for each transaction.
- **Product**: The name of the product purchased.
- **Quantity Ordered**: Number of units ordered for the product.
- **Price Each**: Price of a single unit of the product.
- **Order Date**: The date and time when the order was placed.
- **Purchase Address**: The address where the product was delivered.

The data can be used to:
- Analyze product demand and customer purchasing patterns.
- Identify trends in sales volume over time.
- Understand geographic patterns in purchases.


 <a id="two"></a>
## 2. Importing Packages and Loading Data

We read and merge CSV files for Amazon's monthly sales for 2009 into a final CSV file. We then examine the dataset so we can familiarise ourselves with the uploaded data.

In [1]:
import pandas as pd
import os
import glob

def csvfilesReadMerge(folder_path, new_file_name):
    """
    This function takes multiple CSV files with the same schema from a given folder 
    and combines them into a single large CSV file.
    
    Parameters:
    folder_path (str): Absolute path of the folder where the CSV files are located.
    new_file_name (str): The name of the final CSV file that will be created after merging.
    """
    try:
        # Search for all CSV files in the provided folder path
        files = os.path.join(folder_path, "*.csv")
        list_of_files = glob.glob(files)  # Get list of CSV files
        
        if not list_of_files:
            print("No CSV files found in the specified folder.")
            return
        
        # Read and concatenate all CSV files
        df = pd.concat(map(pd.read_csv, list_of_files), ignore_index=True)
        
        # Save the merged DataFrame to a new CSV file
        df.to_csv(new_file_name, index=False)
        
        # Display success message
        print(f"Successfully created the file '{new_file_name}' at location:\n{os.getcwd()}")
    
    except Exception as e:
        print(f"An error occurred: {e}")

C:\Users\44475\anaconda3\lib\site-packages\pandas\core\arrays\masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.5' currently installed).
  from pandas.core import (


In [2]:
csvfilesReadMerge("C:/Users/44475/Downloads/RawData","AllYearSales.csv")

Successfully created the file 'AllYearSales.csv' at location:
C:\Users\44475


In [3]:
import pandas as pd
df=pd.read_csv("AllYearSales.csv")
df

Order ID                     Product Quantity Ordered Price Each  \
0        176558        USB-C Charging Cable                2      11.95   
1           NaN                         NaN              NaN        NaN   
2        176559  Bose SoundSport Headphones                1      99.99   
3        176560                Google Phone                1        600   
4        176560            Wired Headphones                1      11.99   
...         ...                         ...              ...        ...   
186845   259353      AAA Batteries (4-pack)                3       2.99   
186846   259354                      iPhone                1        700   
186847   259355                      iPhone                1        700   
186848   259356      34in Ultrawide Monitor                1     379.99   
186849   259357        USB-C Charging Cable                1      11.95   

            Order Date                         Purchase Address  
0       04/19/19 08:46             917 1st St, Dallas, TX 75001  
1                  NaN                                      NaN  
2       04/07/19 22:30        682 Chestnut St, Boston, MA 02215  
3       04/12/19 14:38     669 Spruce St, Los Angeles, CA 90001  
4       04/12/19 14:38     669 Spruce St, Los Angeles, CA 90001  
...                ...                                      ...  
186845  09/17/19 20:56   840 Highland St, Los Angeles, CA 90001  
186846  09/01/19 16:00  216 Dogwood St, San Francisco, CA 94016  
186847  09/23/19 07:39     220 12th St, San Francisco, CA 94016  
186848  09/19/19 17:30   511 Forest St, San Francisco, CA 94016  
186849  09/30/19 00:18   250 Meadow St, San Francisco, CA 94016  

[186850 rows x 6 columns]

#### **Basic Information About the Dataset**

The dataset contains `186,850` rows and `6` columns. Below is a summary of the data structure:


In [4]:
# function to perform basic eda operations like null value count, datatype of columns ,shape of data
def basic_eda(pandas_dataframe_object):
    """
    Docstring : This function gives basic info about the dataframe.
    Information returned by the Function are : 
    Shape, (ColumnName, Non-Null Count, Dtype) using info() function and Null Values in the dataset using isna().sum()
    Parameters 
    pandas_dataframe_object : Variable
    Object of pandas dataframe in which dataset is loaded.
    """
    print(f"Shape of data : {pandas_dataframe_object.shape}\n{'-'*50}") 
    print(f"{pandas_dataframe_object.info()}\n{'-'*50}")
    print(f"Count of null values in columns :\n{pandas_dataframe_object.isna().sum()} \n{'-'*50}")
basic_eda(df) 

Shape of data : (186850, 6)
--------------------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 186850 entries, 0 to 186849
Data columns (total 6 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   Order ID          186305 non-null  object
 1   Product           186305 non-null  object
 2   Quantity Ordered  186305 non-null  object
 3   Price Each        186305 non-null  object
 4   Order Date        186305 non-null  object
 5   Purchase Address  186305 non-null  object
dtypes: object(6)
memory usage: 8.6+ MB
None
--------------------------------------------------
Count of null values in columns :
Order ID            545
Product             545
Quantity Ordered    545
Price Each          545
Order Date          545
Purchase Address    545
dtype: int64 
--------------------------------------------------


#### **Sales Transactions Data Summary (2019)**

This section provides a summary of the dataset, which includes sales transactions from an online electronics store.

**Shape of the Data**

The dataset contains **186,850 rows** and **6 columns**, with each row representing a product sold in a specific transaction.

We observe that there are some rows that are completely null so we drop them from the data.

In [5]:
# In the dataframe there are some rows which are completely null so dropping them
Final_data = df.dropna()

basic_eda(Final_data)

Shape of data : (186305, 6)
--------------------------------------------------
<class 'pandas.core.frame.DataFrame'>
Index: 186305 entries, 0 to 186849
Data columns (total 6 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   Order ID          186305 non-null  object
 1   Product           186305 non-null  object
 2   Quantity Ordered  186305 non-null  object
 3   Price Each        186305 non-null  object
 4   Order Date        186305 non-null  object
 5   Purchase Address  186305 non-null  object
dtypes: object(6)
memory usage: 9.9+ MB
None
--------------------------------------------------
Count of null values in columns :
Order ID            0
Product             0
Quantity Ordered    0
Price Each          0
Order Date          0
Purchase Address    0
dtype: int64 
--------------------------------------------------


#### **Sales Transactions Data Summary (2019)**

This section provides an updated summary of the sales transaction dataset after data cleaning and reveals that we do not have any missing values. The dataset captures sales made at an online electronics store during 2019.

Below we split the `Purchase Address` to isolate `State` and `City` to use for our analysis.

#### **Data Cleaning and Preparation**

In this section, we will clean the dataset by extracting relevant geographic information from the `Purchase Address` column and then removing it from the DataFrame.

**Extracting City and State**

To facilitate geographic analysis, we need to create separate columns for `City` and `State`. We will split the `Purchase Address` column to extract these details.

In [6]:
# For Analysis we need City and State columns
Final_data[[ 'City', 'State']] = Final_data['Purchase Address'].str.split(',', expand=True).loc[:,1:]

C:\Users\44475\AppData\Local\Temp\ipykernel_37684\3596567923.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Final_data[[ 'City', 'State']] = Final_data['Purchase Address'].str.split(',', expand=True).loc[:,1:]
C:\Users\44475\AppData\Local\Temp\ipykernel_37684\3596567923.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Final_data[[ 'City', 'State']] = Final_data['Purchase Address'].str.split(',', expand=True).loc[:,1:]


In [7]:
# Dropping Purchase Address from Dataframe as we have extracted City and State Columns
Final_data = Final_data.drop('Purchase Address' , axis =1)

Final_data.head()

Order ID                     Product Quantity Ordered Price Each  \
0   176558        USB-C Charging Cable                2      11.95   
2   176559  Bose SoundSport Headphones                1      99.99   
3   176560                Google Phone                1        600   
4   176560            Wired Headphones                1      11.99   
5   176561            Wired Headphones                1      11.99   

       Order Date          City      State  
0  04/19/19 08:46        Dallas   TX 75001  
2  04/07/19 22:30        Boston   MA 02215  
3  04/12/19 14:38   Los Angeles   CA 90001  
4  04/12/19 14:38   Los Angeles   CA 90001  
5  04/30/19 09:27   Los Angeles   CA 90001

In [8]:
basic_eda(Final_data)

Shape of data : (186305, 7)
--------------------------------------------------
<class 'pandas.core.frame.DataFrame'>
Index: 186305 entries, 0 to 186849
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   Order ID          186305 non-null  object
 1   Product           186305 non-null  object
 2   Quantity Ordered  186305 non-null  object
 3   Price Each        186305 non-null  object
 4   Order Date        186305 non-null  object
 5   City              185950 non-null  object
 6   State             185950 non-null  object
dtypes: object(7)
memory usage: 11.4+ MB
None
--------------------------------------------------
Count of null values in columns :
Order ID              0
Product               0
Quantity Ordered      0
Price Each            0
Order Date            0
City                355
State               355
dtype: int64 
--------------------------------------------------


From above, we can see some rows in `City` and `State` columns are missing , so we have to drop those rows because they will create problems in our analysis.

In [9]:
Final_data.dropna(inplace=True)

In [10]:
basic_eda(Final_data)

Shape of data : (185950, 7)
--------------------------------------------------
<class 'pandas.core.frame.DataFrame'>
Index: 185950 entries, 0 to 186849
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   Order ID          185950 non-null  object
 1   Product           185950 non-null  object
 2   Quantity Ordered  185950 non-null  object
 3   Price Each        185950 non-null  object
 4   Order Date        185950 non-null  object
 5   City              185950 non-null  object
 6   State             185950 non-null  object
dtypes: object(7)
memory usage: 11.3+ MB
None
--------------------------------------------------
Count of null values in columns :
Order ID            0
Product             0
Quantity Ordered    0
Price Each          0
Order Date          0
City                0
State               0
dtype: int64 
--------------------------------------------------


#### Unique Products in the Dataset

To understand the variety of products sold, we will extract the unique product names from the `Product` column of our dataset.
This list provides an overview of the various products included in our dataset, which can be useful for further analysis.

In [11]:
Final_data['Product'].unique()

array(['USB-C Charging Cable', 'Bose SoundSport Headphones',
       'Google Phone', 'Wired Headphones', 'Macbook Pro Laptop',
       'Lightning Charging Cable', '27in 4K Gaming Monitor',
       'AA Batteries (4-pack)', 'Apple Airpods Headphones',
       'AAA Batteries (4-pack)', 'iPhone', 'Flatscreen TV',
       '27in FHD Monitor', '20in Monitor', 'LG Dryer', 'ThinkPad Laptop',
       'Vareebadd Phone', 'LG Washing Machine', '34in Ultrawide Monitor'],
      dtype=object)

#### **Function: product_category**

This function is designed to categorize products in a DataFrame based on their names. We create the following categories based on our unique products:
- Charging Cable,
- Headphones,
- Phone,
- Laptop,
- Monitor,
- Batteries, 
- TV,
- Dryer and
- Washing Machine

In [12]:
def product_category(dataframe, column_name, product_name_ends_with, category_name):
    """
    Docstring : This function is created 'Category' column in dataframe.
    Parameters  
    dataframe : Variable
    Object of pandas dataframe of dataset
    column_name : str enclosed in double inverted commas ("")
    Name of Column where product_name is given
    product_name_ends_with : str enclosed in double inverted commas ("") 
    Name of product with which it ends which will help in identifing the product
    category_name : str enclosed in double inverted commas ("")
    Category name which will be assigned to the product, if found
    """
    index_list=dataframe[dataframe[column_name].str.endswith(product_name_ends_with)==True].index
    dataframe.loc[index_list,"Category"]= category_name
    print(f"'Category' column created Successfully for Category '{category_name}'")

In [13]:
product_category(Final_data, "Product", "Charging Cable", "Charging Cable")

'Category' column created Successfully for Category 'Charging Cable'


In [14]:
product_category(Final_data, "Product", "Headphones", "Headphones")

'Category' column created Successfully for Category 'Headphones'


In [15]:
product_category(Final_data, "Product", "Phone", "Phone")

'Category' column created Successfully for Category 'Phone'


In [16]:
product_category(Final_data, "Product", "Laptop", "Laptop")

'Category' column created Successfully for Category 'Laptop'


In [17]:
product_category(Final_data, "Product", "Monitor", "Monitor")

'Category' column created Successfully for Category 'Monitor'


In [18]:
product_category(Final_data, "Product", "Batteries (4-pack)", "Batteries")

'Category' column created Successfully for Category 'Batteries'


In [19]:
product_category(Final_data, "Product", "TV", "TV")

'Category' column created Successfully for Category 'TV'


In [20]:
product_category(Final_data, "Product", "Dryer", "Dryer")

'Category' column created Successfully for Category 'Dryer'


In [21]:
product_category(Final_data, "Product", "Washing Machine", "Washing Machine")

'Category' column created Successfully for Category 'Washing Machine'


#### **Updated DataFrame Overview**

After categorizing the products, we can examine the updated DataFrame to see the changes reflected in the new `Category` column.

In [22]:
Final_data.head()

Order ID                     Product Quantity Ordered Price Each  \
0   176558        USB-C Charging Cable                2      11.95   
2   176559  Bose SoundSport Headphones                1      99.99   
3   176560                Google Phone                1        600   
4   176560            Wired Headphones                1      11.99   
5   176561            Wired Headphones                1      11.99   

       Order Date          City      State        Category  
0  04/19/19 08:46        Dallas   TX 75001  Charging Cable  
2  04/07/19 22:30        Boston   MA 02215      Headphones  
3  04/12/19 14:38   Los Angeles   CA 90001           Phone  
4  04/12/19 14:38   Los Angeles   CA 90001      Headphones  
5  04/30/19 09:27   Los Angeles   CA 90001      Headphones

#### **Checking for Missing Values**

To ensure the integrity of our dataset, it is essential to check for any missing values across the columns. This helps us confirm that our data is complete and ready for analysis.

In [23]:
Final_data.isnull().sum()

Order ID            0
Product             0
Quantity Ordered    0
Price Each          0
Order Date          0
City                0
State               0
Category            0
dtype: int64

We save 

In [24]:
Final_data.to_csv('AllYearSales.csv',index=False)